Fix `SkinEditorOverlay` freezing when `ReplayPlayer` screen exits early

rushiiMachine opened the pull request:
This addresses issue #25877

Originally when popping in, the ReplayPlayer was loaded first (if previous screen was MainMenu), and afterwards the SkinEditor component was loaded asynchronously. However, if the ReplayPlayer screen exits quickly (like in the event the beatmap has no objects), the skin editor component has not finished initializing (this is before it was even added to the component tree, so it's still not marked `Visible`), then the screen exiting will cause `OsuGame` to call SetTarget(newScreen) -> setTarget(...) which sees that the cached `skinEditor` is not visible yet, and hides/nulls the field. This is the point where LoadComponentAsync(editor, ...) finishes, and the callback sees that the cached skinEditor field is now different (null) than the one that was loaded, and never adds it to the component tree. This occurrence is unhandled and as such the SkinEditorOverlay never hides itself, consuming all input infinitely.

This PR changes the loading to start loading the ReplayPlayer *after* the SkinEditor has been loaded and added to the component tree. Additionally, this lowers the exit delay for ReplayPlayer and changes the "no hit objects" notification to not be an error since it's a controlled exit.

rushiiMachine committed:
Fix `SkinEditorOverlay` freezing when `ReplayPlayer` screen exits early

Originally when popping in, the ReplayPlayer was loaded first (if previous screen was MainMenu), and afterwards the SkinEditor component was loaded asynchronously. However, if the ReplayPlayer screen exits quickly (like in the event the beatmap has no objects), the skin editor component has not finished initializing (this is before it was even added to the component tree, so it's still not marked `Visible`), then the screen exiting will cause `OsuGame` to call SetTarget(newScreen) -> setTarget(...) which sees that the cached `skinEditor` is not visible yet, and hides/nulls the field. This is the point where LoadComponentAsync(editor, ...) finishes, and the callback sees that the cached skinEditor field is now different (null) than the one that was loaded, and never adds it to the component tree. This occurrence is unhandled and as such the SkinEditorOverlay never hides itself, consuming all input infinitely.

This PR changes the loading to start loading the ReplayPlayer *after* the SkinEditor has been loaded and added to the component tree.
Additionally, this lowers the exit delay for ReplayPlayer and changes the "no hit objects" notification to not be an error since it's a controlled exit.

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -547,7 +547,7 @@ private IBeatmap loadPlayableBeatmap(Mod[] gameplayMods, CancellationToken cance
                 if (playable.HitObjects.Count == 0)
                 {
-                    Logger.Log("Beatmap contains no hit objects!", level: LogLevel.Error);
+                    Logger.Log("Beatmap contains no hit objects!", level: LogLevel.Important);
```
Please change this back. This is an error condition.

rushiiMachine replied:
Why so? It's a handled exit that can be caused by users intentionally (not like an uncaught exception).
I can spot another condition in `Player` that looks like it should actually be shown as an error
https://github.com/ppy/osu/blob/668ce937a8d879ebf982ca4bdf1a8ccb1c6cfe36/osu.Game/Screens/Play/Player.cs#L202-L205

bdach replied:
> I can spot another condition in `Player` that looks like it should actually be shown as an error

sure, it probably should be. point is that it's an unexpected condition and it should be communicated to the user as such. i believe stable also posts an error-ish notification on this same circumstance.

peppy replied:
The game also posts `Important` level, FWIW. One may argue that important is more correct here because it's not the game erroring but the beatmap just being wrong, dunno.

https://github.com/ppy/osu/blob/cc800a18b203e474fc9fe6d43f50c2bc4e63ea03/osu.Game/OsuGame.cs#L1179-L1190

rushiiMachine replied:
> One may argue that important is more correct here because it's not the game erroring but the beatmap just being wrong

Yeah that's what makes more sense to me. 
Do Error level logs not also get reported to sentry? Seems very odd to report an issue not actually caused by the game itself as an error.

bdach replied:
> Don't Error level logs also get reported to sentry?

If there is no associated exception (which there is not in this case) then it will not be.

rushiiMachine replied:
> i believe stable also posts an error-ish notification on this same circumstance.

Stable handles this in an even worse way, just presenting an error of "beatmap could not be loaded"

bdach commented on osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs:
```diff
@@ -316,7 +316,7 @@ protected override void LoadComplete()
 
                 if (!LoadedBeatmapSuccessfully)
-                    Scheduler.AddDelayed(this.Exit, 3000);
+                    Scheduler.AddDelayed(this.Exit, RESULTS_DISPLAY_DELAY);
```
Why reuse a results const? What does one have to do with the other?

bdach replied:
Also why was this shortened to begin with?

rushiiMachine replied:
The delay was way too awkwardly long from a ux perspective, given it's a completely blank screen with only the background. For comparison, when trying to actually play a map with 0 objects, the delay before returning to song select isn't 3s like here either.

bdach reviewed (changes requested):
Test coverage would be useful (if feasible)

bdach commented:
> and the callback sees that the cached skinEditor field is now different (null) than the one that was loaded, and never adds it to the component tree. This occurrence is unhandled and as such the SkinEditorOverlay never hides itself, consuming all input infinitely.

How does reordering the operation fix this?

It sounds to me that the skin editor should be doing something to hide itself in this case rather than remain in a zombie state consuming all input forevermore. The presence of player or whatever other screen seems purely incidental in this.

rushiiMachine commented:
> How does reordering the operations fix this?

Since the SkinEditor is added into the component tree and therefore marked Visible first, in the event the ReplayPlayer screen gets added and then exits afterwards does not cause `setTarget()` to nullify the `skinEditor`, since it's already marked Visisble
This whole loading scheme seems pretty jank and easily fragile though.

> It sounds to me that the skin editor should be doing something to hide itself

Then it would exit back into the main menu, I don't think that would be correct behavior either.

> Test coverage would be useful (if feasible)

Truthfully I have no idea how I would even test this, idea would be welcome

rushiiMachine committed:
Don't reuse results delay const

bdach commented:
What I'm getting from this exchange is that the real problem appears to be that `SetTarget()` expects the skin editor to be loaded. Which generally does not have to be true in _any_ circumstance where that method is called.

So I'd say the real fix is to delay the call until the skin editor is ready to receive it. Probably by plugging in another condition into this check:

https://github.com/ppy/osu/blob/668ce937a8d879ebf982ca4bdf1a8ccb1c6cfe36/osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs#L255-L259

rushiiMachine committed:
SkinEditor lifetime fix & show gameplay

If the SkinEditor was created already but not finished initializing, wait for it to initialize before handling a screen change, which could possibly null the skin editor. Additionally, in the case the skin editor is already loaded but hidden, make sure to show gameplay.

peppy commented on osu.Game/Overlays/SkinEditor/SkinEditorOverlay.cs:
```diff
@@ -316,7 +320,7 @@ protected override void LoadComplete()
                 base.LoadComplete();
 
                 if (!LoadedBeatmapSuccessfully)
-                    Scheduler.AddDelayed(this.Exit, 3000);
```
what's going on here?

rushiiMachine replied:
https://github.com/ppy/osu/pull/26149#discussion_r1437279778

peppy commented:
No test coverage? Seems worthwhile to have, and will help with review.

bdach committed:
Merge branch 'master' into fix-skin-editor-init-fail

bdach commented:
As far as I can tell this fixes the (pretty bad issue) and adding a test is going to be annoying, so I'll just get this in on manual testing.

I'll also keep the irrelevant log severity / transition duration changes because they don't really matter.
